Panic in plan output when string contains "null" with whitespace

<!--
Hi there,

Thank you for opening an issue. Please note that we try to keep the Terraform issue tracker reserved for bug reports and feature requests. For general usage questions, please see: https://www.terraform.io/community.html.

If your issue relates to a specific Terraform provider, please open it in the provider's own repository. The index of providers is at https://github.com/terraform-providers .
-->

### Terraform Version
<!---
Run `terraform -v` to show the version, and paste the result between the ``` marks below.

If you are not running the latest version of Terraform, please try upgrading because your issue may have already been fixed.
-->

```
Terraform v0.12.10
```
Also reproducible on v0.12.11-dev (commit a7af2a40)

### Terraform Configuration Files
<!--
Paste the relevant parts of your Terraform configuration between the ``` marks below.

For large Terraform configs, please use a service like Dropbox and share a link to the ZIP file. For security, you can also encrypt the files using our GPG public key.
-->

```hcl
provider "aws" {
  version = "~>2"
  region  = "us-east-1"
}

resource "aws_s3_bucket" "test" {
  bucket = "null "
}
```
This is a contrived minimal example - the bug is independent of provider and was first observed with the Azure provider.

### Debug Output
<!--
Full debug output can be obtained by running Terraform with the environment variable `TF_LOG=trace`. Please create a GitHub Gist containing the debug output. Please do _not_ paste the debug output in the issue, since debug output is long.

Debug output may contain sensitive information. Please review it before posting publicly, and if you are concerned feel free to encrypt the files using the HashiCorp security public key.
-->

### Crash Output
<!--
If the console output indicates that Terraform crashed, please share a link to a GitHub Gist containing the output of the `crash.log` file.
-->
```
panic: value is null

goroutine 14 [running]:
github.com/zclconf/go-cty/cty.Value.LengthInt(0x22a47a0, 0x3416fd0, 0x0, 0x0, 0x0)
	/opt/teamcity-agent/work/9e329aa031982669/pkg/mod/github.com/zclconf/go-cty@v1.1.0/cty/value_ops.go:874 +0x253
github.com/hashicorp/terraform/command/format.(*blockBodyDiffPrinter).writeValue(0xc0005df8a8, 0x22a4760, 0xc00003c460, 0x1a639e0, 0xc00102cbd0, 0x2b, 0x8)
	/opt/teamcity-agent/work/9e329aa031982669/src/github.com/hashicorp/terraform/command/format/diff.go:509 +0x13a8
github.com/hashicorp/terraform/command/format.(*blockBodyDiffPrinter).writeAttrDiff(0xc0005df8a8, 0xc00051d2e8, 0x6, 0xc000b3d4a0, 0x22a4760, 0xc00003c460, 0x0, 0x0, 0x22a4760, 0xc00003c460, ...)
	/opt/teamcity-agent/work/9e329aa031982669/src/github.com/hashicorp/terraform/command/format/diff.go:255 +0x370
github.com/hashicorp/terraform/command/format.(*blockBodyDiffPrinter).writeBlockBodyDiff(0xc0005df8a8, 0xc000af5570, 0x22a4860, 0xc0004570f8, 0x0, 0x0, 0x22a4860, 0xc000457310, 0x1b08d40, 0xc000a38f90, ...)
	/opt/teamcity-agent/work/9e329aa031982669/src/github.com/hashicorp/terraform/command/format/diff.go:196 +0x5d5
github.com/hashicorp/terraform/command/format.ResourceChange(0xc000bf8200, 0xc0008c9400, 0xc000af5570, 0xc0003a5130, 0xc000a8b860, 0x8e)
	/opt/teamcity-agent/work/9e329aa031982669/src/github.com/hashicorp/terraform/command/format/diff.go:140 +0x5ff
github.com/hashicorp/terraform/backend/local.(*Local).renderPlan(0xc00018a410, 0xc0002540e0, 0xc000457180, 0xc000454080)
	/opt/teamcity-agent/work/9e329aa031982669/src/github.com/hashicorp/terraform/backend/local/backend_plan.go:270 +0x6ad
github.com/hashicorp/terraform/backend/local.(*Local).opPlan(0xc00018a410, 0x22a2ae0, 0xc00026d8c0, 0x22a2ae0, 0xc00026d900, 0xc0002a28c0, 0xc00026d880)
	/opt/teamcity-agent/work/9e329aa031982669/src/github.com/hashicorp/terraform/backend/local/backend_plan.go:165 +0xb9c
github.com/hashicorp/terraform/backend/local.(*Local).Operation.func1(0xc0003a5900, 0xc0003a5910, 0xc0003a5920, 0xc0002a28c0, 0xc00018a410, 0xc00026d880, 0xc0003a58f0, 0x22a2ae0, 0xc00026d8c0, 0x22a2ae0, ...)
	/opt/teamcity-agent/work/9e329aa031982669/src/github.com/hashicorp/terraform/backend/local/backend.go:356 +0x135
created by github.com/hashicorp/terraform/backend/local.(*Local).Operation
	/opt/teamcity-agent/work/9e329aa031982669/src/github.com/hashicorp/terraform/backend/local/backend.go:340 +0x36b
```
[Full log](https://gist.github.com/simonbrady/77c719927fc57c98eab9a33a6ccc450d)

### Expected Behavior
<!--
What should have happened?
-->
Output of `terraform plan` should include the literal string `"null "`.

### Actual Behavior
<!--
What actually happened?
-->
The above crash occurs consistently.

### Steps to Reproduce
<!--
Please list the full steps required to reproduce the issue, for example:
1. `terraform init`
2. `terraform apply`
-->
1. `terraform init`
2. `terraform plan`

### Additional Context
<!--
Are there anything atypical about your situation that we should know? For example: is Terraform running in a wrapper script or in a CI system? Are you passing any unusual command line options or environment variables to opt-in to non-default behavior?
-->

### References
<!--
Are there any other GitHub issues (open or closed) or Pull Requests that should be linked here? For example:

- #6017

-->
- Fix in PR #23102

--- a/command/format/diff.go
+++ b/command/format/diff.go
@@ -502,7 +502,7 @@ func (p *blockBodyDiffPrinter) writeValue(val cty.Value, action plans.Action, in
 				ty, err := ctyjson.ImpliedType(src)
 				// check for the special case of "null", which decodes to nil,
 				// and just allow it to be printed out directly
-				if err == nil && !ty.IsPrimitiveType() && val.AsString() != "null" {
+				if err == nil && !ty.IsPrimitiveType() && strings.TrimSpace(val.AsString()) != "null" {
 					jv, err := ctyjson.Unmarshal(src, ty)
 					if err == nil {
 						p.buf.WriteString("jsonencode(")

--- a/command/format/diff_test.go
+++ b/command/format/diff_test.go
@@ -52,6 +52,26 @@ func TestResourceChange_primitiveTypes(t *testing.T) {
     }
 `,
 		},
+		"creation (null string with extra whitespace)": {
+			Action: plans.Create,
+			Mode:   addrs.ManagedResourceMode,
+			Before: cty.NullVal(cty.EmptyObject),
+			After: cty.ObjectVal(map[string]cty.Value{
+				"string": cty.StringVal("null "),
+			}),
+			Schema: &configschema.Block{
+				Attributes: map[string]*configschema.Attribute{
+					"string": {Type: cty.String, Optional: true},
+				},
+			},
+			RequiredReplace: cty.NewPathSet(),
+			Tainted:         false,
+			ExpectedOutput: `  # test_instance.example will be created
+  + resource "test_instance" "example" {
+      + string = "null "
+    }
+`,
+		},
 		"deletion": {
 			Action: plans.Delete,
 			Mode:   addrs.ManagedResourceMode,
@@ -205,6 +225,37 @@ new line
       ~ more_lines = <<~EOT
             original
           + new line
+        EOT
+    }
+`,
+		},
+		"addition of multi-line string field": {
+			Action: plans.Update,
+			Mode:   addrs.ManagedResourceMode,
+			Before: cty.ObjectVal(map[string]cty.Value{
+				"id":         cty.StringVal("i-02ae66f368e8518a9"),
+				"more_lines": cty.NullVal(cty.String),
+			}),
+			After: cty.ObjectVal(map[string]cty.Value{
+				"id": cty.UnknownVal(cty.String),
+				"more_lines": cty.StringVal(`original
+new line
+`),
+			}),
+			Schema: &configschema.Block{
+				Attributes: map[string]*configschema.Attribute{
+					"id":         {Type: cty.String, Optional: true, Computed: true},
+					"more_lines": {Type: cty.String, Optional: true},
+				},
+			},
+			RequiredReplace: cty.NewPathSet(),
+			Tainted:         false,
+			ExpectedOutput: `  # test_instance.example will be updated in-place
+  ~ resource "test_instance" "example" {
+      ~ id         = "i-02ae66f368e8518a9" -> (known after apply)
+      + more_lines = <<~EOT
+            original
+            new line
         EOT
     }
 `,
@@ -857,11 +908,11 @@ func TestResourceChange_JSON(t *testing.T) {
 			Mode:   addrs.ManagedResourceMode,
 			Before: cty.ObjectVal(map[string]cty.Value{
 				"id":         cty.StringVal("i-02ae66f368e8518a9"),
-				"json_field": cty.StringVal(`[{"one": "111"}, {"two": "222"}]`),
+				"json_field": cty.StringVal(`[{"one": "111"}, {"two": "222"}, {"three": "333"}]`),
 			}),
 			After: cty.ObjectVal(map[string]cty.Value{
 				"id":         cty.UnknownVal(cty.String),
-				"json_field": cty.StringVal(`[{"one": "111"}]`),
+				"json_field": cty.StringVal(`[{"one": "111"}, {"three": "333"}]`),
 			}),
 			Schema: &configschema.Block{
 				Attributes: map[string]*configschema.Attribute{
@@ -881,6 +932,9 @@ func TestResourceChange_JSON(t *testing.T) {
                 },
               - {
                   - two = "222"
+                },
+                {
+                    three = "333"
                 },
             ]
         )

